File: src/recordnote/transcriber.py
"""Speech-to-text transcription module using OpenAI Whisper."""

import tempfile
from pathlib import Path
from typing import Dict, Optional

import whisper
from whisper.model import Whisper


class SpeechTranscriber:
    """Speech transcriber using OpenAI Whisper for Japanese audio."""

    def __init__(self, model_size: str = "base") -> None:
        """Initialize the speech transcriber.

        Args:
            model_size: Whisper model size (tiny, base, small, medium, large)
        """
        self.model_size = model_size
        self._model: Optional[Whisper] = None

    def load_model(self) -> None:
        """Load the Whisper model. Called automatically when needed."""
        if self._model is None:
            print(f"Loading Whisper model: {self.model_size}")
            self._model = whisper.load_model(self.model_size)

    def transcribe_file(self, audio_file_path: Path) -> Dict[str, str]:
        """Transcribe audio file to text.

        Args:
            audio_file_path: Path to the audio file

        Returns:
            Dictionary containing transcribed text and language info
        """
        if not audio_file_path.exists():
            raise FileNotFoundError(f"Audio file not found: {audio_file_path}")

        self.load_model()
        assert self._model is not None

        # Transcribe with Japanese language specified
        result = self._model.transcribe(
            str(audio_file_path), language="ja", verbose=False
        )

        return {
            "text": result["text"].strip(),
            "language": result["language"],
            "segments": [
                {
                    "start": segment["start"],
                    "end": segment["end"],
                    "text": segment["text"].strip(),
                }
                for segment in result["segments"]
            ],
        }

    def transcribe_bytes(self, audio_bytes: bytes) -> Dict[str, str]:
        """Transcribe audio from bytes data.

        Args:
            audio_bytes: Audio data as bytes (WAV format)

        Returns:
            Dictionary containing transcribed text and language info
        """
        # Create temporary file for audio bytes
        with tempfile.NamedTemporaryFile(suffix=".wav", delete=False) as temp_file:
            temp_file.write(audio_bytes)
            temp_path = Path(temp_file.name)

        try:
            return self.transcribe_file(temp_path)
        finally:
            # Clean up temporary file
            temp_path.unlink(missing_ok=True)

    def get_model_info(self) -> Dict[str, str]:
        """Get information about the loaded model.

        Returns:
            Dictionary with model information
        """
        return {
            "model_size": self.model_size,
            "loaded": self._model is not None,
        }

File: recordnote-rust/src/speech/transcriber.rs
use anyhow::{Context, Result};
use hound::WavReader;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscriptionSegment {
    pub start: f64,
    pub end: f64,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscriptionResult {
    pub text: String,
    pub language: String,
    pub segments: Vec<TranscriptionSegment>,
}

pub struct WhisperTranscriber {
    model_size: String,
    loaded: bool,
}

impl WhisperTranscriber {
    pub fn new(model_size: String) -> Result<Self> {        
        Ok(Self {
            model_size,
            loaded: false,
        })
    }

    pub fn default() -> Result<Self> {
        Self::new("base".to_string())
    }

    pub async fn load_model(&mut self) -> Result<()> {
        if self.loaded {
            return Ok(()); // Already loaded
        }

        log::info!("Loading Whisper model: {}", self.model_size);
        
        // Placeholder for actual model loading
        // In a real implementation, you would:
        // 1. Download the Whisper model if not cached
        // 2. Load the model weights using whisper.cpp bindings or similar
        // 3. Initialize the tokenizer
        
        // Simulate loading delay
        tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
        
        self.loaded = true;
        
        log::info!("Model loaded successfully");
        Ok(())
    }

    pub async fn transcribe_file(&mut self, audio_path: &Path) -> Result<TranscriptionResult> {
        if !audio_path.exists() {
            return Err(anyhow::anyhow!("Audio file not found: {:?}", audio_path));
        }

        self.load_model().await?;

        // Load and preprocess audio
        let audio_data = self.load_audio_file(audio_path)?;
        
        // For now, this is a placeholder implementation
        // In a real implementation, you would:
        // 1. Preprocess the audio (resample to 16kHz, normalize, etc.)
        // 2. Run the Whisper model inference
        // 3. Decode the output tokens to text
        // 4. Apply timestamp alignment
        
        // Placeholder result for Japanese text
        let result = TranscriptionResult {
            text: "これは音声認識の結果です。実際の実装では、Whisperモデルを使用して音声をテキストに変換します。".to_string(),
            language: "ja".to_string(),
            segments: vec![
                TranscriptionSegment {
                    start: 0.0,
                    end: 3.0,
                    text: "これは音声認識の結果です。".to_string(),
                },
                TranscriptionSegment {
                    start: 3.0,
                    end: 8.0,
                    text: "実際の実装では、Whisperモデルを使用して音声をテキストに変換します。".to_string(),
                },
            ],
        };

        Ok(result)
    }

    pub async fn transcribe_bytes(&mut self, audio_bytes: &[u8]) -> Result<TranscriptionResult> {
        // Create a temporary file from bytes
        let temp_file = tempfile::NamedTempFile::with_suffix(".wav")?;
        std::fs::write(temp_file.path(), audio_bytes)?;
        
        self.transcribe_file(temp_file.path()).await
    }

    pub fn get_model_info(&self) -> HashMap<String, serde_json::Value> {
        let mut info = HashMap::new();
        info.insert("model_size".to_string(), serde_json::Value::String(self.model_size.clone()));
        info.insert("loaded".to_string(), serde_json::Value::Bool(self.loaded));
        info.insert("device".to_string(), serde_json::Value::String("cpu".to_string()));
        info
    }

    fn load_audio_file(&self, path: &Path) -> Result<Vec<f32>> {
        let mut reader = WavReader::open(path)?;
        let spec = reader.spec();
        
        if spec.sample_rate != 16000 {
            log::warn!("Audio sample rate is {}, expected 16000. Resampling may be needed.", spec.sample_rate);
        }

        let samples: Result<Vec<f32>, _> = match spec.sample_format {
            hound::SampleFormat::Float => {
                reader.samples::<f32>().collect()
            }
            hound::SampleFormat::Int => {
                reader.samples::<i32>()
                    .map(|s| s.map(|s| s as f32 / i32::MAX as f32))
                    .collect()
            }
        };

        let samples = samples?;
        
        // Convert stereo to mono if needed
        let mono_samples = if spec.channels == 2 {
            samples
                .chunks(2)
                .map(|chunk| (chunk[0] + chunk[1]) / 2.0)
                .collect()
        } else {
            samples
        };

        Ok(mono_samples)
    }
}

// Helper function to resample audio (simplified implementation)
fn resample_audio(samples: Vec<f32>, from_rate: u32, to_rate: u32) -> Vec<f32> {
    if from_rate == to_rate {
        return samples;
    }
    
    // Simplified linear interpolation resampling
    let ratio = from_rate as f64 / to_rate as f64;
    let new_length = (samples.len() as f64 / ratio) as usize;
    
    let mut resampled = Vec::with_capacity(new_length);
    
    for i in 0..new_length {
        let src_index = (i as f64 * ratio) as usize;
        if src_index < samples.len() {
            resampled.push(samples[src_index]);
        }
    }
    
    resampled
}